FileInfo, file_info, and link_info

tynanbe opened the pull request:
- ~~Renames `file.is_file` to `file.exists`~~
- Removes `file.is_file`
- Adds `FileInfo`, `FileType`, and `Access` types to the `gleam/erlang/file` module
- Adds `file_info`, `link_info`, `file_exists`, `link_exists`, and `is_regular` functions to the `gleam/erlang/file` module
- Updates `file.is_directory` to return a `Result(Bool, Reason)`
- Updates all doc examples in the `gleam/erlang/file` module to use the triple backtick syntax

lpil commented on src/gleam/erlang/file.gleam:
```diff
@@ -123,19 +123,32 @@ pub type Reason {
   "filelib" "is_dir"
 
-/// Returns true if path refers to a file, otherwise false.
+/// Returns true if path refers to a regular file, otherwise false.
```
What does `regular` mean? Not a word I've come across before when talking about files. Would `is_file` be a clearer name?

tynanbe replied:
For e.g. the `test` command available in most shells

```sh
test
  -d FILE
      Returns true if FILE is a directory.
  -e FILE
      Returns true if FILE exists.
  -f FILE
      Returns true if FILE is a regular file.
```

The main issue here is Erlang's choice of names for these functions.
`filelib:is_file` is `test -e`
`filelib:is_regular` is `test -f`

I'm open to other names, but I think it shouldn't be confusing. Currently, I'd say it's confusing.

`file.is_regular` seems fine to me. `file.is_file` would potentially be confusing again if e.g. someone overlooks `file.exists`.

At first I went with `is_regular_file`, but then thought it was redundant given the module name.

tynanbe replied:
More specifically, as far as I know, `regular` means it's not a directory, link, or any other "special" type of file.

lpil replied:
That's interesting. I'm still not sure what it means, could we add the definition of `regular` to the docs?

tynanbe replied:
According to Elixir docs and a bit of testing, a "regular" file is one that isn't a "directory, socket, symbolic link, named pipe, or device file."

tynanbe commented:
After further digging, testing, and discussion on the Discord server, I refactored the code and introduced some new types and functions. Changelog and docs unfinished; I'd appreciate some feedback about the current approach before I write the docs~

tynanbe commented:
@lpil following up on this to see what you think about it before I proceed with docs. Thanks~

lpil commented on src/gleam/erlang/file.gleam:
```diff
@@ -111,32 +111,146 @@ pub type Reason {
+    size: Int,
+    file_type: FileType,
+    access: Access,
+    atime: Int,
+    mtime: Int,
+    ctime: Int,
```
Do you think it'd be worth spelling these out in full? We tend to avoid abbreviations

tynanbe replied:
I'd tend to agree; however, we have at least one issue here.
As per https://www.erlang.org/doc/man/file.html#read_file_info-1

## `atime`

The last time the file was read.

## `mtime`

The last time the file was written.

## `ctime`

The interpretation of this time field depends on the operating system. On Unix, it is the last time the file or the inode was changed. In Windows, it is the create time.

## Fullnames
- `access_time` or `accessed_time`
- `modify_time` or `modified_time` or `modification_time`
- `???_time` (possibly `meta_or_create_time` or `change_or_create_time` or the Gleam-like `crange_time` — which also has the same number of letters as `access_time` and `modify_time` :stuck_out_tongue_winking_eye:)

Perhaps `atime`, `mtime`, and `ctime` are familiar enough and clearer as is.

lpil replied:
Fair enough, short names it is!

lpil commented on src/gleam/erlang/file.gleam:
````diff
@@ -111,32 +111,146 @@ pub type Reason {
+/// ```gleam
+/// ```
+///
+pub external fn read_file_info(path: String) -> Result(FileInfo, Reason) =
````
Perhaps just `file_info`?

lpil replied:
info returning FileInfo sounds good to me

tynanbe replied:
I forgot about `link_info`, so I went with `file_info` to match. Also changed `exists` to `file_exists` and added `link_exists`.

lpil reviewed (commented):
Looks fantastic! Thank you. Left 2 small questions but really liking the look of this

lpil reviewed (approved):
Brilliant work, thank you very much!
